File: XpmSample/XpmSample.Android.Native/xpmsamplelib/src/main/java/com/xamarin/xpmsamplelib/XpmSampleLib.java
package com.xamarin.xpmsamplelib;

import android.content.Context;
import android.util.Log;

import com.xamarin.platformmessaging.MessageChannel;

public class XpmSampleLib {
    static MessageChannel _channel;
    static final String Tag = "XpmSampleLib";
    static final String ChannelId = "XpmSampleLib";
    static final String GetTextMessageId = "get_text";
    static final String LogTextMessageId = "log_text";
    static final String TextLoggedMessageId = "text_logged";

    public static void start(Context context) {
        _channel = new MessageChannel(ChannelId);

        _channel.setResponseHandler(GetTextMessageId, (parameters) -> {
            return "Hello from Java";
        });

        _channel.setHandler(LogTextMessageId, (parameters) -> {
            String arg = (parameters != null && parameters.length == 1) ? (String)parameters[0] : null;

            if (arg != null) {
                Log.i(Tag, arg);
            }
            else {
                Log.i(Tag, "No text provided");
            }

            try
            {
                _channel.send(TextLoggedMessageId, arg);
            }
            catch (Exception ex) {
                Log.e(Tag, ex.getLocalizedMessage());
            }
        });
    }

    public static void stop() {
        if (_channel == null) {
            return;
        }

        try {
            _channel.clearResponseHandler(GetTextMessageId);
            _channel.clearHandler(LogTextMessageId);
        }
        catch (Exception ex) {
            Log.e(Tag, ex.getLocalizedMessage());
        }

        _channel = null;
    }
}

File: XpmSample/XpmSample.iOS.Native/XpmSampleLib/XpmSampleLib.swift
//
//  XpmSampleLib.swift
//  XpmSampleLib
//
//  Created by Mike Parker on 29/01/2021.
//

import Foundation
import XamarinPlatformMessaging

@objc(XpmSampleLib)
public class XpmSampleLib : NSObject {
    static var _channel: MessageChannel?
    static let _channelId = "XpmSampleLib"
    static let  _getTextMessageId = "get_text";
    static let  _logTextMessageId = "log_text";
    static let  _textLoggedMessageId = "text_logged";
    
    @objc(start)
    public static func start() {
        _channel = MessageChannel(channelId: _channelId)
        
        if let channel = _channel {
            channel.setResponseHandler(forMessageId: _getTextMessageId) { (parameters) in
               return "Hello from Swift" as NSObject
            }
            
            channel.setHandler(forMessageId: _logTextMessageId) { (parameters) in
                do {
                    if let args = parameters, let arg = args.first as? NSString {
                        NSLog(arg.description)
                        try channel.send(messageId: _textLoggedMessageId, parameters: arg)
                    }
                    else {
                        NSLog("No text provided")
                        try channel.send(messageId: _textLoggedMessageId, parameters: nil)
                    }
                } catch {
                    NSLog(error.localizedDescription)
                }
            }
        }
    }
    
    @objc(stop)
    public static func stop() {
        guard let channel = _channel else {return}
        
        do {
            try channel.clearResponseHandler(forMessageId: _getTextMessageId)
            try channel.clearHandler(forMessageId: _logTextMessageId)
        }
        catch {
            NSLog(error.localizedDescription)
        }
        
        _channel = nil
    }
}
